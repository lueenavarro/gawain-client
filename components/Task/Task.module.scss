@use "styles/colors";

.task {
  height: 30px;
  position: relative;
  box-sizing: border-box;

  &--completed {
    color: colors.$gray;
    text-decoration: line-through;
  }

  &__wrapper {
    position: relative;
    min-height: 28.5px;
    padding: 0 20px;
    box-sizing: border-box;
    overflow: hidden;
    display: flex;
  }

  &__text {
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    min-width: 0;
    margin-right: 10px;
    padding: 5px 0;
  }

  &:hover {
    z-index: 10;

    .task__wrapper {
      background-color: colors.$lightgray;
      overflow: visible;
    }

    .task__text {
      overflow: visible;
      white-space: unset;
      word-break: break-word;
    }

    .task__delete {
      display: flex;
      align-items: center;
      justify-content: center;
      font-weight: bold;
    }
  }

  &__delete {
    display: none;
    margin-left: auto;
    cursor: pointer;
    width: 25px;

    &:hover {
      color: colors.$red;
    }
  }
}
