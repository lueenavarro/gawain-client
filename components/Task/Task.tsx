import React from "react";
import PropTypes from "prop-types";

import styles from "./Task.module.scss";

const Task = (props) => {
  const { task, onRemove, onComplete, snapshot } = props;

  const taskBgClass = () =>
    `${styles.task} ${snapshot.isDragging ? styles["task--active"] : ""}`;
  const taskClass = () => `${styles["task__text"]} ${task.completed ? styles["task--completed"] : ""}`;

  return (
    <div
      className={taskBgClass()}
      onClick={() => onComplete(task._id, !task.completed)}
      data-testid="taskBg"
    >
      <div className={styles["task__wrapper"]}>
      <span className={taskClass()} data-testid="task">
        {task.task}
      </span>
      <div
        className={styles["task__delete"]}
        onClick={(e) => {e.stopPropagation(); onRemove(task._id)}}
        data-testid="remove"
      >
        x
      </div>
      </div>
    </div>
  );
};

Task.propTypes = {
  task: PropTypes.any.isRequired,
  onRemove: PropTypes.func.isRequired,
  onComplete: PropTypes.func.isRequired,
  snapshot: PropTypes.any,
};

export default Task;
