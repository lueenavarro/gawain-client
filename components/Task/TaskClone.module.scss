.task__wrapper {
  background-color: pink;
}
