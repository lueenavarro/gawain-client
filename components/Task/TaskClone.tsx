import React from "react";

import styles from "./Task.module.scss";
import cloneStyles from "./TaskClone.module.scss";

export const TaskClone = ({ item: task }) => {
  const taskClass = () =>
    `${styles["task__text"]} ${
      task.completed ? styles["task--completed"] : ""
    }`;

  return (
    <div className={styles.task}>
      <div
        className={`${styles["task__wrapper"]} ${cloneStyles["task__wrapper"]}`}
      >
        <span className={taskClass()} data-testid="task">
          {task.task}
        </span>
      </div>
    </div>
  );
};
