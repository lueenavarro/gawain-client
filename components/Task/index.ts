export * from "./Task";
export * from "./TaskClone";
