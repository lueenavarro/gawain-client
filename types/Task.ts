export interface ITask {
  _id: string;
  task: string;
  completed: boolean;
}
