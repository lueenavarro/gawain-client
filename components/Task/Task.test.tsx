import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";

import { cleanup, render } from "test/react";
import Task from "./Task";

const mockOnRemove = jest.fn((_id: string) => {});
const mockOnComplete = jest.fn(() => {});
const fakeTask = { _id: "1234", task: "Go To Market" };
const fakeSnapshot = { isDragging: false };

afterEach(cleanup);

test("render without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(
    <Task
      task={fakeTask}
      onRemove={mockOnRemove}
      snapshot={fakeSnapshot}
      onComplete={mockOnComplete}
    />,
    div
  );
});

test("render task correctly", () => {
  const { getByTestId } = render(
    <Task
      task={fakeTask}
      onRemove={mockOnRemove}
      snapshot={fakeSnapshot}
      onComplete={mockOnComplete}
    />
  );
  expect(getByTestId("task")).toHaveTextContent("Go To Market");
});

test("set active class", () => {
  fakeSnapshot.isDragging = true;
  const { getByTestId } = render(
    <Task
      task={fakeTask}
      onRemove={mockOnRemove}
      snapshot={fakeSnapshot}
      onComplete={mockOnComplete}
    />
  );
  expect(getByTestId("taskBg").className).toBe("task task--active");
});

test("call onRemove when x is clicked", () => {
  const { getByTestId } = render(
    <Task
      task={fakeTask}
      onRemove={mockOnRemove}
      snapshot={fakeSnapshot}
      onComplete={mockOnComplete}
    />
  );
  getByTestId("remove").click();
  expect(mockOnRemove).toBeCalledWith("1234");
});

test("snapshot matches", () => {
  const tree = renderer
    .create(
      <Task
        task={fakeTask}
        onRemove={mockOnRemove}
        snapshot={fakeSnapshot}
        onComplete={mockOnComplete}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
