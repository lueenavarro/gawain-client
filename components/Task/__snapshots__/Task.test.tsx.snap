// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot matches 1`] = `
<div
  className="task task--active"
  data-testid="taskBg"
  onClick={[Function]}
>
  <div
    className="task__wrapper"
  >
    <span
      className="task__text "
      data-testid="task"
    >
      Go To Market
    </span>
    <div
      className="task__delete"
      data-testid="remove"
      onClick={[Function]}
    >
      x
    </div>
  </div>
</div>
`;
